frontend/model/notifications/vuexModule.js

```diff
@@ -0,0 +1,104 @@
+import type {
+  Notification,
+  VuexModuleContext
+} from './types.flow.js'
+
+import sbp from '~/shared/sbp.js'
+import './selectors.js'
+import { age, isNew, isOlder } from './utils.js'
+import * as keys from './mutationKeys.js'
+
+const actions = {
+  async login (context: VuexModuleContext, user: Object) {
+    const notifications = await sbp('gi.db/notifications/load', user.username)
+
+    if (notifications && notifications.length) {
+      context.commit(keys.SET_NOTIFICATIONS, notifications)
+    }
+  },
+  async logout (context: VuexModuleContext) {
+    if (context.rootState.loggedIn) {
+      // Make sure to save the notifications *before* clearing them.
+      await sbp('gi.notifications/save')
+      context.commit(keys.REMOVE_ALL_NOTIFICATIONS)
+    }
+  }
+}
+
+const getters = {
+  newNotifications (state, getters) {
+    return state.filter(item => isNew(item))
+  },
+
+  newNotificationCount (state, getters) {
+    return getters.newNotifications.length
+  },
+
+  notificationCount (state, getters) {
+    return state.length
+  },
+
+  olderNotifications (state, getters) {
+    return state.filter(item => isOlder(item))
+  },
+
+  olderNotificationCount (state, getters) {
+    return getters.olderNotifications.length
+  },
+
+  unreadNotifications (state, getters) {
+    return state.filter(item => !item.read)
+  },
+
+  unreadNotificationCount (state, getters) {
+    return getters.unreadNotifications.length
+  }
+}
+
+const mutations = {
+  [keys.ADD_NOTIFICATION] (state, notification: Notification) {
+    if (state.includes(notification)) {
```

Comment:
Unsure I follow the entire logic here, but if I might say, maybe these shouldn't be 'hard' errors, because they might for some reason come repeated or out of order?

Reply:
These errors should normally not occur since `emit` always creates a new object, with a new timestamp.

Should such error nonetheless happen for some reason, then it's likely the symptom of a bug in the API or incorrect usage thereof somewhere. For example, manual notification programming using plain JS rather than the SBP API.

In such case, I think it's better to be notified of any issue immediately so as to be able to fix it. Otherwise, other developers might start relying on unsupported behavior, possibly leading to more complicated issues.

Reply:
Yes @snowteamer I think @Silver-IT is correct, `state.includes` will only compare by reference:

```js
> var a = {a: 4}
undefined
> b = []
[]
> b.push(a)
1
> b.includes({a: 4})
false
```

Reply:
@greg That's intended. This check is meant to prevent duplicate references from ever being included in the array. However distinct objects with equal keys and values are allowed for now. Please see https://github.com/okTurtles/group-income/pull/1138#discussion_r804097925